### PR TITLE
common/build-style/python3-pep517.sh: ignore tests in temp destdir

### PR DESCRIPTION
**Pull request by icp1994**
#### Testing the changes
- I tested the changes in this PR: **YES**

An edge case I came across today. On rare cases, packages include tests in their built wheel. While testing with these wheels extracted in a temporary directory, it conflicts with the tests in the source tree as pytest cannot run testfiless with same name in different directories (ref: https://github.com/pytest-dev/pytest/issues/3151).

See [this run](https://github.com/void-linux/void-packages/actions/runs/5021644643/jobs/9004230165) as an example. The same package passes checks correctly in this PR.

### CODE REVIEW

**Comment by tornaria** [comment]
This PR fixes the first time check is run, but not the second and subsequent runs.

In fact, the wheel will be installed into a different testdir each time (and not removed afterwards) and this only ignores the current one.

Suggestions:
 - use `testdir=$(mktemp -d)` so the testdir is outside the wrksrc tree.
 - use `--ignore=${wrksrc}/tmp` so we ignore all installations.

**Comment by tornaria** [comment]
LGTM now. Just FTR this  merge confilcts with #44071 but both PR are necessary.

**Comment by icp1994** [comment]
Yeah I'll rebase once that's merged into master.

**Comment by ahesford** [comment]
This seems reasonable. Using `mktemp` to create directories outside of `$wrksrc` is undesirable because we might potentially leave artifacts that aren't cleaned up by `xbps-src clean`. Sure, it will be in `/tmp`, but it's still better to be sure it gets cleaned up after the build.

**Comment by icp1994** [comment]
Where would you put the `rm` call though? If the `pytest` call fails, it won't get to that.

**Comment by ahesford** [comment]
That's okay, because failing tests will already abort a build and any subsequent attempt probably ought to start from a clean masterdir for proper behavior anyway. Cleaning up at the end of `do_check` after a successful test should be good enough. Leaving the temporary directory around after a failed test also lets users inspect the installation for defects that might have triggered the error.

**Comment by icp1994** [comment]
Oh, okay. I thought the norm was if the build is successful, one should be able to run `./xbps-src check` multiple times on failed tests without running `./xbps-src clean`

**Comment by ahesford** [comment]
Sure, it's reasonable to repeatedly run `check` to work out test issues, but when it comes time to build and install the final package, I definitely recommend a clean build first.

File: common/build-style/python3-pep517.sh [hunk 1]
```diff
@@ -24,7 +24,10 @@ do_check() {
 		${make_install_args} ${make_install_target:-dist/*.whl}
 
 	PATH="${testdir}/usr/bin:${PATH}" PYTHONPATH="${testdir}/${py3_sitelib}" \
-		${make_check_pre} pytest3 ${testjobs} ${make_check_args} ${make_check_target}
+		${make_check_pre} pytest3 --ignore="${wrksrc}/tmp" ${testjobs} \
+		${make_check_args} ${make_check_target}
+
+	rm -r "${wrksrc}/tmp"
```
[the remark this answers is not in the document]
Comment (ahesford) [reply common/build-style/python3-pep517.sh:30]: Alternatively, we could do:
```suggestion
	rm -r "${testdir}"
	rmdir tmp >/dev/null 2>&1 || true
```

[the remark this answers is not in the document]
Comment (tornaria) [reply common/build-style/python3-pep517.sh:30]: I agree with removing only what was created in this run (i.e. just `${tmpdir}`).

Also, in terms of avoiding conflicts, maybe the tmpdir should be something more "unique". Perhaps `${wrksrc}/.xbps-tmpdir/$(data +%s)` which also has the advantage that pytest will not recurse hidden directories, so the `--ignore ..` doesn't seem to be necessary.

**Comment by tornaria** [comment]
While we are at this, I was going to suggest using a venv instead of `--destdir` + `PYTHONPATH` trick. Indeed, I found some packages that fail when `PYTHONPATH` is messed upon.

The gist of it would be something like:
```
local testdir="${wrksrc}/.xbps-tmpdir/$(date +%s)"
python3 -m venv --system-site-packages --without-pip ${testdir}
${testdir}/bin/python3 -m installer ${make_install_args} ${make_install_target:-dist/*.whl}

PATH=${testdir}/bin:$PATH ${make_check_pre} python3 -m \
    pytest ${testjobs} ${make_check_args} ${make_check_target}
```

This is needed to fix check for `python3-coverage`, see: https://github.com/void-linux/void-packages/pull/44160/commits/4366b1ccf70c3dd531416aec2f75f95f5ad5827c.

I also needed it to fix check for `python3-pytest-cov`, however in that case another hack was necessary so I would still need to rewrite `do_check()`. See: https://github.com/void-linux/void-packages/pull/44160/commits/65eb1057479d8dff616e56dcafacf3d587657170.

File: common/build-style/python3-pep517.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (icp1994) [reply common/build-style/python3-pep517.sh:30]: I'm a bit confused about the need to remove the `$testdir` after a successful run. What's the issue if we keep it there?

And if we are adding `/.xbps-tmpdir` for uniqueness, I don't see why we also need to include another sub-directory under it. I only added the epoch subdirs for the extremely rare chance if a package actually ships a `/tmp` dirrctory under `$wrksrc`.

[the remark this answers is not in the document]
Comment (tornaria) [reply common/build-style/python3-pep517.sh:30]: Fine with me in principle. Although I'd be wary about what happens when there is an old testdir lying around from a previous check run. I don' t know if `python -m installer` will do the right thing.

[the remark this answers is not in the document]
Comment (ahesford) [reply common/build-style/python3-pep517.sh:30]: I guess this is irrelevant because the wheel was already built before the tmpdir was created, and the installer should really only be unpacking the wheel. I do prefer @tornaria's suggestion to give the directory a hidden, XBPS-specific name that is extremely unlikely to conflict with package contents. Maybe try changing `testdir` to `.xbps-testdir` (which is more meaningful than `.xbps-tempdir`) and see if that solves your problem without needing to ignore tests, and then skip the removal.

**Comment by ahesford** [comment]
I'm not convinced we should move to a venv for testing. Half of your two test cases require further adjustment (symlinking a package into the venv) anyway. Adjusting the Python path is a supported mechanism, so the failure of those two packages probably reflects some inappropriate assumptions in those packages.

File: common/build-style/python3-pep517.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (icp1994) [reply common/build-style/python3-pep517.sh]: After some testing, the python installer doesn't override files so moved the deletion logic at top. And removed the now redundant epoch subdir & ignore flag.

**Comment by tornaria** [comment]
This LGTM, let's get this in, we can figure out whether switching or not to testing using a venv later (cf #44268 where I needed a venv).

File: common/build-style/python3-pep517.sh [hunk 2]
```diff
@@ -19,7 +19,9 @@ do_check() {
 		testjobs="-n $XBPS_MAKEJOBS"
 	fi
 
-	local testdir="${wrksrc}/tmp/$(date +%s)"
+	local testdir="${wrksrc}/.xbps-testdir"
+	rm -rf "${testdir}"
```
Comment (ahesford) [inline common/build-style/python3-pep517.sh:23]: The benefit of cleaning up the test directory *after* we've used it is that we (attempt to) restore the source tree to the state it was in when the function was first called.

The benefit of cleaning up the directory before it would normally exist seems dubious. If we agree there is no need to clean up the test after a successful run, I think changing the directory name as you've done is sufficient.

Comment (ahesford) [reply common/build-style/python3-pep517.sh:23]: I just ran into the reason for the cleanup: attempting to re-run a check after a failure will cause `python3 -m installer` to fail unless the test directory is clean. Still, I think the idea of purging a (typically nonexistent) directory just to avoid issues with re-running tests is strange. At this point, I guess I'd rather see
```sh
local testdir="${wrksrc}/.xbps-testdir/$(date +s)"
```
and skip cleanup altogether. This should solve the problem of `pytest` trying to traverse into `tmp` without requiring a manual exclude, keeps our tests in a more-likely-unique subdirectory that should not conflict with any package contents, and avoid strange pre-clean steps. As a bonus, if you *do* run multiple tests to correct a failure, you can always do something like
```sh
diff -r .xbps-testdir/<failed> .xbps-testdir/<succeeeded>
```
after the fact to confirm what, if anything, has changed in the installed wheel.

Comment (icp1994) [reply common/build-style/python3-pep517.sh]: Ayyy, feels like I'm running in circles :laughing: This should be the Final.Final.Final.patch
